File: fibonacci/fibPython/fib.py
# Iterative Fibonacci
# Return the nth number using a loop.
# It starts from 0 and 1 and adds the previous two numbers until it reaches n.
def fib_iterative(n):
    if n <= 1:
        return n

    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b 
    return b

# Recursive Fibonacci
# It uses the formula: fib(n) = fib(n-1) + fib(n-2) until it reaches the base case.
def fib_recursive(n):
    if n <= 1:
        return n
    return fib_recursive(n - 1) + fib_recursive(n - 2)

# Dynamic Programming Fibonacci
# Return the nth number using a list to store previous results
# Avoids recalculating the same values and is faster than simple recursion for large n.
def fib_dynamic(n):
    if n <= 1:
        return n

    fib_numbers = [0, 1]
    for i in range(2, n + 1):
        fib_numbers.append(fib_numbers[i - 1] + fib_numbers[i - 2])
    return fib_numbers[n]

File: fibonacci/fibC/fib.c
#include "fib.h"
#include <stdio.h>

// global variable to count operations
unsigned long long FIB_OPS = 0;

void fib_reset_ops(void) {
    FIB_OPS = 0;
}

// Iterative 
// uses a simple loop to get the nth number

long long fib_iterative(int n){
if (n<= 1)
return n;

long long a = 0;
long long b = 1;
long long next = 0;

for(int i =2; i<= n; i++){
    next = a + b; // adds the two previous number 
    a = b;   // shift a forward
    b = next; // shift b forward
    FIB_OPS++;  // counts total number of additions 
}

return b; 

}

// Recursive
// calls itself again and again until it reaches the base case 

long long fib_recursive(int n){
    if (n <= 1)
        return n;

    FIB_OPS++;
    return fib_recursive(n - 1) + fib_recursive(n - 2);

}

// Dynamic Programming Fibonacci
// This version stores the results, no need to repeat steps
long long fib_dynamic(int n) {
    if (n <= 1)
        return n;

    long long fib[100]; // small array to hold values
    fib[0] = 0;
    fib[1] = 1;

    for (int i = 2; i <= n; i++) {
        fib[i] = fib[i - 1] + fib[i - 2];
        FIB_OPS++;
    }

    return fib[n];


}
